src/spawn.c

```diff
@@ -0,0 +1,1250 @@
+ *     - as a consequence, an unqouted # at the start of an argument comments to the end of line
+ *  Windows:
+ *     - a quoted program name must be entirely inside the quotes. It's hard to believe, but
+ *	   "C:\Foo\Bar".pdf is actually executed as "C:\Foo\Bar.exe", and .pdf is completely
+ *	   ignored, not even passed as an argument
+ *     - the standard Windows quoting rules are used: double quote is escaped with slash, and
+ *       any literal slashes before a double quote must be duplicated.
+ *
+ *  @param command_line the command line to check.
+ *  @param execute whether to check, using @c g_find_program_in_path(), if the program specified
+ *         in the command line exists and is executable.
+ *  @param error return location for error.
+ *
+ *  @return @c TRUE on success, @c FALSE if an error was set.
+ **/
+gboolean spawn_check_command(const gchar *command_line, gboolean execute, GError **error)
+{
+	gchar *program;
+
+#ifdef G_OS_WIN32
+	gboolean open_quote = FALSE;
+	const gchar *s, *arguments;
+
+	g_return_val_if_fail(command_line != NULL, FALSE);
+
+	while (*command_line && strchr(CL_BLANKS, *command_line))
+		command_line++;
+
+	if (!*command_line)
+	{
+		/* TL note: from glib */
+		g_set_error(error, G_SHELL_ERROR, G_SHELL_ERROR_EMPTY_STRING,
+			_("Text was empty (or contained only whitespace)"));
+		return FALSE;
+	}
+
+	if (*command_line == '"')
+	{
+		command_line++;
+		/* Windows allows "foo.exe, but we may have extra arguments */
+		if ((s = strchr(command_line, '"')) == NULL)
+		{
+			/* TL note: from glib */
+			g_set_error(error, G_SHELL_ERROR, G_SHELL_ERROR_BAD_QUOTING,
+				_("Text ended before matching quote was found for %c."
+				  " (The text was '%s')"), '"', command_line);
+			return FALSE;
+		}
+
+		if (!strchr(CL_BLANKS, s[1]))  /* strchr() catches s[1] == '\0' */
+		{
+			g_set_error(error, G_SHELL_ERROR, G_SHELL_ERROR_BAD_QUOTING,
+				_("A quoted Windows program name must be entirely inside the quotes"));
+			return FALSE;
+		}
+	}
+	else
+	{
+		const gchar *quote = strchr(command_line, '"');
+
+		/* strchr() catches *s == '\0', and the for body is empty */
+		for (s = command_line; !strchr(CL_BLANKS, *s); s++);
+
+		if (quote && quote < s)
+		{
+			g_set_error(error, G_SHELL_ERROR, G_SHELL_ERROR_BAD_QUOTING,
+				_("A quoted Windows program name must be entirely inside the quotes"));
+			return FALSE;
+		}
+	}
+
+	program = g_strndup(command_line, s - command_line);
+	arguments = s + (*s == '"');
+
+	for (s = arguments; *s; s++)
+	{
+		if (*s == '"')
+		{
+			const char *slash;
+
+			for (slash = s; slash > arguments && slash[-1] == '\\'; slash--);
+			if ((s - slash) % 2 == 0)
+				open_quote ^= TRUE;
+		}
+	}
+
+	if (open_quote)
+	{
+		/* TL note: from glib */
+		g_set_error(error, G_SHELL_ERROR, G_SHELL_ERROR_BAD_QUOTING,
+			_("Text ended before matching quote was found for %c."
+			  " (The text was '%s')"), '"', command_line);
+		return FALSE;
```

Comment:
Fixed. I can save s, or use different variables, but it's easier to just free it.